lib/driver-cli.js

```diff
@@ -0,0 +1,250 @@
+      // dereference the actual npm package ('appiupm-xcuitest-driver'), so
+      // check it exists and get the correct package
+      const knownNames = Object.keys(KNOWN_DRIVERS);
+      if (!_.includes(knownNames, name)) {
+        const msg = `Could not resolve driver; are you sure it's in the list ` +
+                    `of supported drivers? ${JSON.stringify(knownNames)}`;
+        throw new Error(msg);
+      }
+      pkgName = KNOWN_DRIVERS[name];
+      // given that we'll use the install type in the driver json, store it as
+      // 'npm' now
+      installType = INSTALL_TYPE_NPM;
+    }
+
+    driverData = await installViaNpm({appiumHome, json, driver, pkgName, pkgVer});
+  }
+
+  const {driverName} = driverData;
+  delete driverData.driverName;
+
+  if (_.includes(Object.keys(INSTALLED_DRIVERS), driverName)) {
+    throw new Error(`A driver named '${driverName}' is already installed. ` +
+                    `Did you mean to update? 'appium driver update'. See ` +
+                    `installed drivers with 'appium driver list --installed'.`);
+  }
+
+  driverData.installType = installType;
+  driverData.installSpec = installSpec;
+  await writeInstalledDriver(appiumHome, driverName, driverData);
+
+  // refresh the installed driver manifest
+  INSTALLED_DRIVERS = await validateDriversYaml(appiumHome, (msg) => log(json, msg));
+
+  // log info for the user
+  log(json, `Driver ${driverName}@${driverData.version} successfully installed`.green);
+  log(json, `- automationName: ${driverData.automationName.green}`);
+  log(json, `- platformNames: ${JSON.stringify(driverData.platformNames).green}`);
+
+  return INSTALLED_DRIVERS;
+}
+
+async function installViaNpm ({appiumHome, json, driver, pkgName, pkgVer}) {
+  let installPath = pkgName;
+  if (!pkgName) {
+    pkgName = driver;
+    installPath = path.posix.basename(pkgName);
+  }
+  const npmSpec = `${pkgName}${pkgVer ? '@' + pkgVer : ''}`;
+  const specMsg = npmSpec === driver ? '' : ` using NPM install spec '${npmSpec}'`;
+  const msg = `Installing '${driver}'${specMsg}`;
+  try {
+    const pkgJsonData = await spinWith(json, msg, async () => (
+      await installPackage(path.resolve(appiumHome, installPath), pkgName, pkgVer)
+    ));
+    const driverData = getDriverFields(pkgJsonData);
+    driverData.installPath = installPath;
+    return driverData;
+  } catch (err) {
+    throw new Error(`Encountered an error when installing package: ${err.message}`);
+  }
+}
+
+function getDriverFields (pkgJsonData) {
+  if (!pkgJsonData.appium) {
+    throw new Error(`Installed driver did not have an 'appium' section in its ` +
+                    `package.json file as expected`);
+  }
+
+  const {appium, name, version} = pkgJsonData;
+  const {driverName, automationName, platformNames, mainClass} = appium;
+
+  const reqFields = ['driverName', 'automationName', 'platformNames', 'mainClass'];
+  const missingFields = reqFields.reduce((acc, field) => (
```

Comment:
could we use lodash _.difference helper instead?

Reply:
i don't think so, we're not trying to find the difference between two arrays. we're trying to find which keys from a short list don't exist on an object. there's probably a way to recast this problem in terms of _.difference but i don't think it would be clearer than reduce.